### ISSUE TITLE
Upgrading incompatible issue

### ISSUE THREAD

**Issue by SElab2019** [google/gson#1566]
Hi,

The following code snippets throw an incompatible issue when I try to upgrade Gson from 2.2.2  to 2.8.5.

```
public class TestGson {
	
    public static <T> String convertObjectToString(T t, Class<T> clazz) {
        Gson gson = new GsonBuilder().create();
        try {
            return gson.getAdapter(clazz).toJson(t);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return "";
    }
}
```

It works well before, but It throws an error after upgrading:
```
exception java.io.IOException is never thrown in body of corresponding try statement
```

Thanks a lot!

**Comment by Marcono1234** [comment]
In case this is still relevant, could you please test this with the latest Gson version and also provide a complete code example?

**Comment by jvirtanen** [comment]
This actually looks like an improvement in Gson between 2.2.2 and 2.8.5: in Gson 2.8.5, `TypeAdapter<T>#toJson(T value)` is no longer declared as throwing `IOException`. This was implemented in 032847976c0cba7c131e95aa6608882987f1eca8 and released in Gson 2.4.

**Comment by Marcono1234** [comment]
Ah thank you for the clarification, I thought the author of this issue meant the error message occurred during runtime, but it was actually a compiler error.

I guess adding back `throws IOException` now would not be very useful and would also again break source compatibility. Though the assumption that no `IOException` can occur is actually incorrect (#2172).

Also, let's better not look at the signature `fromJson(String) throws IOException` from a consistency perspective 🙈
